sli/slicontrol.cc

```diff
@@ -1680,54 +1685,34 @@ TimeFunction::execute( SLIInterpreter* i ) const
   i->OStack.push_move( tmp );
 }
 
-/*BeginDocumentation
-Name: sleep_i - suspends proces for n seconds
-Synopsis:  n sleep_i -> -
-Description:
-Calls the POSIX select() function.
-SeeAlso: clock, usertime, tic, toc
-*/
 
 void
-Sleep_iFunction::execute( SLIInterpreter* i ) const
+Sleep_dFunction::execute( SLIInterpreter* i ) const
 {
   i->assert_stack_load( 1 );
 
-  const long sec = static_cast< long >( i->OStack.pick( 0 ) );
-  const long usec = 0;
-  struct timeval tv = { sec, usec };
-  if ( sec > 0 )
+  const double t = static_cast< double >( i->OStack.pick( 0 ) );
+
+  if ( t < 0 )
   {
-    select( 0, 0, 0, 0, &tv );
+    throw BadParameterValue( "t >= 0 required." );
   }
 
-  i->OStack.pop();
-  i->EStack.pop();
-}
-
-/*BeginDocumentation
-Name: sleep_i - suspends proces for x seconds
-Synopsis:  x sleep_d -> -
-Description:
-Calls the POSIX select() function.
-SeeAlso: clock, usertime, tic, toc
-*/
-
-void
-Sleep_dFunction::execute( SLIInterpreter* i ) const
-{
-  i->assert_stack_load( 1 );
-
-  const long sec = 0;
-  const long usec = static_cast< long >(
-    static_cast< double >( i->OStack.pick( 0 ) ) * 1000000. );
-
-  struct timeval tv = { sec, usec };
-  if ( usec > 0 )
+  if ( t > std::numeric_limits< int >::max() )
   {
-    select( 0, 0, 0, 0, &tv );
+    throw BadParameterValue( String::compose(
+      "t < %1s required.", std::numeric_limits< int >::max() ) );
   }
 
+  // split into second and microsecond part
+  // limited to 32-bit signed int, see #973.
+  const int t_sec = static_cast< int >( t );
+  const int t_musec = static_cast< int >( ( t - t_sec ) * 1e6 );
+
+  // on some systems, select may modify tv, therefore, it cannot be const
+  struct timeval tv = { t_sec, t_musec };
+  select( 0, 0, 0, 0, &tv );
```

Comment:
On second thought, I find it a bit weird to use [`select`](https://en.wikipedia.org/wiki/Select_(Unix)) here instead of just
```C++
#include <unistd.h>
...
usleep(microseconds);
```
Can you please change (or at least comment on) this? Thanks!